src/chunked/encoder.rs

```diff
@@ -3,250 +3,57 @@ use std::pin::Pin;
-    ///
-    /// # Format
-    ///
-    /// Each "chunk" uses the following encoding:
-    ///
-    /// ```txt
-    /// 1. {byte length of `data` as hex}\r\n
-    /// 2. {data}\r\n
-    /// ```
-    ///
-    /// A chunk stream is finalized by appending the following:
-    ///
-    /// ```txt
-    /// 1. 0\r\n
-    /// 2. {trailing header}\r\n (can be repeated)
-    /// 3. \r\n
-    /// ```
-    pub(crate) fn encode(
-        &mut self,
-        res: &mut Response,
-        cx: &mut Context<'_>,
-        buf: &mut [u8],
-    ) -> Poll<io::Result<usize>> {
-        self.bytes_written = 0;
-        let res = self.run(res, cx, buf);
-        log::trace!("ChunkedEncoder {} bytes written", self.bytes_written);
-        res
-    }
-
-    /// Execute the right method for the current state.
-    fn run(
-        &mut self,
-        res: &mut Response,
-        cx: &mut Context<'_>,
-        buf: &mut [u8],
-    ) -> Poll<io::Result<usize>> {
-        match self.state {
-            State::Start => self.dispatch(State::EncodeChunks, res, cx, buf),
-            State::EncodeChunks => self.encode_chunks(res, cx, buf),
-            State::EndOfChunks => self.encode_chunks_eos(res, cx, buf),
-            State::ReceiveTrailers => self.receive_trailers(res, cx, buf),
-            State::EncodeTrailers => self.encode_trailers(res, cx, buf),
-            State::EndOfStream => self.encode_eos(res, cx, buf),
-            State::End => Poll::Ready(Ok(self.bytes_written)),
+            reader,
+            done: false,
         }
     }
+}
 
-    /// Switch the internal state to a new state.
-    fn dispatch(
-        &mut self,
-        state: State,
-        res: &mut Response,
-        cx: &mut Context<'_>,
-        buf: &mut [u8],
-    ) -> Poll<io::Result<usize>> {
-        use State::*;
-        log::trace!("ChunkedEncoder state: {:?} -> {:?}", self.state, state);
-
-        #[cfg(debug_assertions)]
-        match self.state {
-            Start => assert!(matches!(state, EncodeChunks)),
-            EncodeChunks => assert!(matches!(state, EndOfChunks)),
-            EndOfChunks => assert!(matches!(state, ReceiveTrailers)),
-            ReceiveTrailers => assert!(matches!(state, EncodeTrailers | EndOfStream)),
-            EncodeTrailers => assert!(matches!(state, EndOfStream)),
-            EndOfStream => assert!(matches!(state, End)),
-            End => panic!("No state transitions allowed after the ChunkedEncoder has ended"),
-        }
-
-        self.state = state;
-        self.run(res, cx, buf)
+fn max_bytes_to_read(buf_len: usize) -> usize {
+    if buf_len < 6 {
+        panic!("buffers of length {} are too small for this implementation. if this is a problem for you, please open an issue", buf_len);
     }
+    let max_bytes_of_hex_framing = // the maximum number of bytes that the hex representation of remaining bytes might take
+                (((buf_len - 5) as f64).log2() / 4f64).floor() as usize;
+    buf_len - 5 - max_bytes_of_hex_framing
```

Comment:
I found this part a bit hard to read; we can smooth out the formatting somewhat by moving parts around:
```suggestion
    // the maximum number of bytes that the hex representation of remaining bytes might take
    let max_bytes_of_hex_framing = (((buf_len - 5) as f64).log2() / 4f64).floor();
    buf_len - 5 - (max_bytes_of_hex_framing  as usize)
```